### PR TITLE
Add some more ergonomics to instance syncing along with sqlite example

### PR DESCRIPTION
**Pull request by erlendvollset**
- Add backoff utilitiy
- Add callback-based method for subscribing to a given dms query
- Add docs along with example on how to sync to local sqlite

### CODE REVIEW

File: docs/source/data_modeling.rst [hunk 1]
```diff
@@ -126,6 +126,96 @@ Query instances
...
+                    (node.space, node.external_id)
+                    for node in result["nodes"]
+                    if node.deleted_time is not None
+                ]
+                # Updates must be done in the same transaction as persisting the cursor.
+                # A transaction is implicitly started by sqlite here.
+                connection.executemany(
+                    "INSERT INTO instance VALUES (?, ?, ?)", inserts
+                )
+                connection.executemany(
```
Comment (alexbrasetvik) [inline docs/source/data_modeling.rst:202]: Delete must happen before insert. The same `space, external_id` may appear as several tombstones and then a new instance, which must result in the instance being saved.

Comment (erlendvollset) [reply docs/source/data_modeling.rst:202]: Ah right. TY 🙏

File: cognite/client/data_classes/data_modeling/instances.py [hunk 2]
```diff
@@ -809,3 +813,11 @@ class InstancesDeleteResult:
 
     nodes: list[NodeId]
     edges: list[EdgeId]
+
+
+@dataclass
+class SubscriptionContext:
+    _canceled: bool = False
```
Comment (erlendvollset) [inline cognite/client/data_classes/data_modeling/instances.py:820]: Note to self: add some heartbeat timestamps like last_successful_sync and last_successful_callback, to make it easy to check for progress.

File: cognite/client/_api/data_modeling/instances.py [hunk 3]
```diff
@@ -394,6 +414,80 @@ def delete(
...
+                    if any(len(instances) > 0 for instances in result.data.values()):
+                        callback(result)
+                        subscription_context.last_successful_callback = datetime.now(tz=timezone.utc)
+                        # only progress the cursor if the callback executed successfully
+                        cursors = result.cursors
+                        backoff.reset()
+                        continue
+                    else:
+                        # No data was returned, so we back off before we try again.
+                        time.sleep(next(backoff))
```
Comment (alexbrasetvik) [inline cognite/client/_api/data_modeling/instances.py:480]: I think polling delay/interval and error backoff should be separate concepts.

If there's a poll interval, I'd delay the first one with `random.uniform(0, poll_interval)` to smear out a bit, but let the subsequent ones go with a consistent delay.

If `self.sync` throws, we shouldn't need to re-query before the next attempt.

File: docs/source/data_modeling.rst [hunk 4]
```diff
@@ -126,6 +126,102 @@ Query instances
...
+    def sync_space_to_sqlite(
+        db_name: str, space_to_sync: str
+    ) -> SubscriptionContext:
+        with sqlite_connection(db_name) as connection:
+            connection.execute(
+                "CREATE TABLE IF NOT EXISTS instance (space TEXT, external_id TEXT, data JSON)"
+            )
+            connection.execute(
+                "CREATE UNIQUE INDEX IF NOT EXISTS space_extid_idx ON instance (space, external_id)"
+            )
```
Comment (alexbrasetvik) [inline docs/source/data_modeling.rst:167]:
```suggestion
            connection.execute(
                "CREATE TABLE IF NOT EXISTS instance (space TEXT, external_id TEXT, data JSON, PRIMARY KEY(space, external_id))"
            )
```

File: docs/source/data_modeling.rst [hunk 5]
```diff
@@ -126,6 +126,102 @@ Query instances
...
+
+
+    def sqlite_connection(db_name: str) -> sqlite3.Connection:
+        return sqlite3.connect(db_name, check_same_thread=False)
+
+
+    def sync_space_to_sqlite(
+        db_name: str, space_to_sync: str
+    ) -> SubscriptionContext:
+        with sqlite_connection(db_name) as connection:
```
Comment (alexbrasetvik) [inline docs/source/data_modeling.rst:161]: Maybe split up into clearer setup vs continuous sync functions?

**Review by alexbrasetvik — CHANGES_REQUESTED** [review]
Looks generally good. I think the error vs polling delays should be separate.

The SQLite example can split up setup vs continuous sync a bit, so we can more clearly excessively comment the latter.

File: cognite/client/_api/data_modeling/instances.py (hunk 3 above)
Comment (alexbrasetvik) [reply cognite/client/_api/data_modeling/instances.py:480]: Sorry, I meant if `callback` throws.

**Review by arnels — APPROVED** [review]
LGTM

**Review by alexbrasetvik — APPROVED** [review]
LGTM.

Let's follow up with future PRs that make the SDK logic not log full traces for intermittent errors like connectivity blips, 429s, etc., but this looks like a good start.

**Pull request merged**